### PR TITLE
Fix #57: Compile selectors at css parsing time.

### LINKED ISSUE #57: Incorrect parsing of styles
#### Steps to reproduce:
`(new ExtendedCss(':contains(/[\\w]{9,}/) { display:none!important; }')).apply()`

#### Expected:
No error is printed in the console.

#### Actual:
```
Extended selector is invalid: /)
Extended selector is invalid: :contains(/[\w]
```

### PR DESCRIPTION
**Pull request by seanl-adg**

### CODE REVIEW

File: lib/extended-css-parser.js [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+                lastTokens.length = l - 1;
+            }
+            selector = selector.trim();
+
+            // Start of style declaration parse loop 
+            var styles = Object.create(null);
+            var match, matchPos, matched;
+            nextIndex++; // Move the pointer to the start of style declaration.
+            while (true) {
+                untilClosingBracket: {
```
Comment (ameshkov) [inline lib/extended-css-parser.js]: The code is too messy. Extract `untilClosingBracket` into a function.

File: lib/extended-css-parser.js [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+            if (lastTokens[l - 1].type === ' ') {
+                lastTokens.length = l - 1;
+            }
+            selector = selector.trim();
+
+            // Start of style declaration parse loop 
+            var styles = Object.create(null);
+            var match, matchPos, matched;
+            nextIndex++; // Move the pointer to the start of style declaration.
+            while (true) {
```
Comment (ameshkov) [inline lib/extended-css-parser.js]: Avoid `while (true)` loops at any cost. There's always a chance of an infinite loop.

If you can't find a better way, use `for (i < SOME_HUGE_NUMBER)` and throw an error if you didn't get the result in the end.

File: lib/extended-css-parser.js [hunk 3]
```diff
@@ -77,130 +77,131 @@ var ExtendedCssParser = (function() { // jshint ignore:line
...
+
+                this.nextIndex++; // Move the pointer to the start of style declaration.
 
-                // Matched a closing bracket - cut out matched portion
-                // from cssText and re-run the outer loop.
-                reNonWhitespace.lastIndex = matchPos + 1;
-                match = reNonWhitespace.exec(cssText);
+                var styleMap = this.parseNextStyle();
+
+                results.push({
```
Comment (ameshkov) [inline lib/extended-css-parser.js]: Empty styleMap is not handled

Comment (ameshkov) [reply lib/extended-css-parser.js]: Worth a unit-test mbe?

Comment (seanl-adg) [reply lib/extended-css-parser.js]: > Empty styleMap is not handled

Comment (ameshkov) [reply lib/extended-css-parser.js]: > Do we need any special treatment?

Nvm, I thought that empty style signals about a mistake. I can see that it's not the case actually.

**Pull request merged**